### PR TITLE
fix(synthetic-shadow): duplicate event listeners on template and host are not discarded

### PR DESCRIPTION
**Pull request by Sumitjg**
Duplicate event listeners on template and host are not discarded.

## GUS work item
W-8928394

### CODE REVIEW

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 1]
```diff
@@ -266,6 +271,16 @@ function shouldInvokeShadowRootListener(event: Event): boolean {
     return isCurrentTargetSlotted || targetHost === currentTargetHost;
 }
 
+function isDuplicate(currentTarget: EventTarget, type: String, listener: EventListener) {
```
Comment (pmdartus) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts:274]:
```suggestion
function isDuplicateListener(currentTarget: EventTarget, type: String, listener: EventListener) {
```

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 2]
```diff
@@ -282,9 +297,14 @@ export function addCustomElementEventListener(
         }
     }
     // TODO [#1824]: Lift this restriction on the option parameter
-    if (isFunction(listener)) {
+    if (isFunction(listener) && !isDuplicate(this, type, listener)) {
         const wrappedListener = getWrappedCustomElementListener(listener);
         attachDOMListener(this, type, wrappedListener);
+
+        const listenerToEventsMap = elementToListenerAndEventsMap.get(this) ?? new WeakMap();
```
Comment (pmdartus) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts:304]: This subroutine should be extracted.

File: packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts [hunk 3]
```diff
@@ -63,6 +63,11 @@ interface ListenerMap {
     [key: string]: WrappedListener[];
 }
 
+const elementToListenerAndEventsMap: WeakMap<
```
Comment (pmdartus) [inline packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts:66]: The `Set<string>` has to be cleaned up when an event listener is manually removed. I am suspecting that currently you can't add, remove and add again the same event listener on a given event target.

Comment (Sumitjg) [reply packages/@lwc/synthetic-shadow/src/faux-shadow/events.ts:66]: yeah, nice catch

**Pull request closed without merging**